src/engineer/nowcast.py

```diff
@@ -0,0 +1,124 @@
+
+        e.g. if I set the `pred_months = 11`
+
+        `x_target_variable` = 11 timesteps
+        `x_non_target_variable` = 12 timesteps (`pred_months + 1`).
+
+        We overcome this by creating an extra timestep with all nan values
+        in the `x_dataset`. This way the `x_dataset` contains the `y_dataset`
+        timestep but the `target_variable` is an array of all `np.nan` for that
+        target timestep. This prevents model leakage.
+        """
+        if expected_length is not None:
+            warnings.warn("Expected length must be None\
+                          for the `nowcast` experiment. Forcing to\
+                          None")
+            expected_length = None
+
+        print(f"Generating data for year: {year}, target month: {target_month}")
+
+        # get the test datetime
+        max_date = date(year, target_month, calendar.monthrange(year, target_month)[-1])
+        mx_year, mx_month, max_train_date = minus_months(
+            year, target_month, diff_months=1
+        )
+        _, _, min_date = minus_months(mx_year, mx_month, diff_months=pred_months)
+
+        # convert to numpy datetime
+        min_date_np = np.datetime64(str(min_date))
+        max_date_np = np.datetime64(str(max_date))
+        max_train_date_np = np.datetime64(str(max_train_date))
+
+        print(
+            f"Max date: {str(max_date)}, max input date: {str(max_train_date)}, "
+            f"min input date: {str(min_date)}"
+        )
+
+        # boolean array indexing the TARGET VARIABLE timestamps to filter `ds`
+        x_target = (ds.time.values > min_date_np) & (
+            ds.time.values <= max_train_date_np
+        )
+        y_target = (ds.time.values > max_train_date_np) & (
+            ds.time.values <= max_date_np
+        )
+
+        # boolean array indexing the other variables
+        x_non_target = (ds.time.values > min_date_np) & (ds.time.values <= max_date_np)
+
+        # only expect ONE y timestamp
+        if sum(y_target) != 1:
+            print(f"Wrong number of y values! Expected 1, got {sum(y_target)};\
+            returning None")
+            return None, cast(date, max_train_date)
+
+        # create the target dataset `y_dataset` & the `x_non_target_dataset`
+        y_dataset = ds.isel(time=y_target)[target_variable].to_dataset(
+            name=target_variable
+        )
+        x_non_target_dataset = ds.drop(target_variable).sel(time=x_non_target)
+
+        # create the x_target_dataset with all nans at target time
+        nan_target_variable = self.make_nan_dataset(y_dataset)
+        x_target_dataset = (
+            ds[target_variable].isel(time=x_target).to_dataset(name=target_variable)
+        )
+        x_target_dataset = x_target_dataset.merge(nan_target_variable)
+
+        # merge the x_non_target_dataset + x_target_dataset -> x_dataset
+        x_dataset = x_non_target_dataset.merge(x_target_dataset)
+
+        if x_dataset.time.size != pred_months + 1:
```

Comment:
This assumes the preprocessors were run with monthly granularity. We probably want to compare this to `expected_length`, not `pred_months`

Reply:
so now the choice of `expected_length = None` as default looks off that's my bad!

does `pred_months` work with non-month timesteps because it's filtering with time instead of number of steps?

Reply:
i now understand why its so important